src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs

```diff
@@ -117,6 +117,63 @@ internal static class CSharpDiagnosticFilter
             return d.WithReportDiagnostic(reportAction);
         }
 
+        internal static ReportDiagnostic FilterAndGetReportDiagnostic(
+            DiagnosticDescriptor d,
+            int warningLevelOption,
+            NullableContextOptions nullableOption,
+            ReportDiagnostic generalDiagnosticOption,
+            IDictionary<string, ReportDiagnostic> specificDiagnosticOptions,
+            SyntaxTreeOptionsProvider? syntaxTreeOptions,
+            CancellationToken cancellationToken)
+        {
+            if (d == null)
+            {
+                return ReportDiagnostic.Suppress;
+            }
+            else if (d.IsNotConfigurable())
+            {
+                if (d.IsEnabledByDefault)
+                {
+                    // Enabled NotConfigurable should always be reported as it is.
+                    return DiagnosticDescriptor.MapSeverityToReport(d.DefaultSeverity);
+                }
+                else
+                {
+                    // Disabled NotConfigurable should never be reported.
+                    return ReportDiagnostic.Suppress;
+                }
+            }
+            else if (d.DefaultSeverity == InternalDiagnosticSeverity.Void)
+            {
+                return ReportDiagnostic.Suppress;
+            }
+
+            var reportAction = GetDiagnosticReport(
```

Comment:
To illustrate why the logic duplication makes me uncomfortable: the special handling for `WRN_ALinkWarn` was removed compared to `Filter` method above. Is that an intentional change? Is it observable?

Reply:
That was intentional, the DiagnosticDescriptor being passed in from DiagnosticDescriptor.GetEffectiveSeverity would never hit that code. But, I see your point, this is a bit fragile. Let me try your other suggestion again and see how it pans out.